kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala

```diff
@@ -119,6 +119,19 @@ class KyuubiSessionImpl(
     engineLastAlive = System.currentTimeMillis()
   }
 
+  def reLaunchEngine(): Unit = handleSessionException {
+    withDiscoveryClient(sessionConf) { discoveryClient =>
+      engine.deregister(
+        discoveryClient,
+        engine.getOrCreate(discoveryClient, launchEngineOp.getOperationLog))
+    }
+    _client.closeSession()
+    engineLaunched = false
+    launchEngineOp = sessionManager.operationManager
+      .newLaunchEngineOperation(this, shouldRunAsync = false)
+    runOperation(launchEngineOp)
```

Comment:
This will re-execute `openEngineSession` and we need to consider whether some states are as expected.

+ `openedTime` will be reset: `sessionEvent.openedTime = System.currentTimeMillis()`
+ do we need to update `engineLastAlive`?

Reply:
fixed `engineLastAlive `, `EngineRef#builder`.
I think `openedTime ` means engine openedTime, so we don't need to update `openedTime`.And i changed the `org.apache.kyuubi.events.KyuubiSessionEvent#openedTime` comment.

Reply:
You shall need to wrap this method with `_client.withLockAcquired`.

Reply:
in case that other operations access the `_client` during restarting engine.

Reply:
We may not be able to use the `lock` of `_client` because `_client` will be assigned a new object after this method.

Reply:
> in case that other operations access the `_client` during restarting engine.

This seems difficult to avoid

Reply:
> Or we need to add `ReadWriteLock` for the `client` method of `KyuubiSessionImpl`.

submit `Operation` from multiple threads will meet problem , I will fix it.